### PR TITLE
[Feature] Shadows with Hamming matrix

### PR DESCRIPTION
**Pull request by chMoussa**
Solves #43 which introduce a method to compute shadows for the case where we have a number of shots per unitary superior to 1.

### CODE REVIEW

File: docs/measurements/index.md [hunk 1]
```diff
@@ -91,10 +91,10 @@ Along the same lines as the example before, estimating the expectation value usi
 
 ```python exec="on" source="material-block" session="measurements" result="json"
 # Classical shadows are defined up to some accuracy and confidence.
-from qadence_protocols.measurements.utils_shadow import number_of_samples
+from qadence_protocols.measurements.utils_shadow.data_acquisition import number_of_samples
 
-shadow_options = {"accuracy": 0.1, "confidence": 0.1}
-N, K = number_of_samples(observable, **shadow_options)
+shadow_options = {"accuracy": 0.1, "confidence": 0.1, "n_shots": 1}
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]: Is it necessary to specify the number of shots here ?

File: docs/measurements/index.md [hunk 2]
```diff
@@ -103,6 +103,11 @@ estimated_values_shadow = shadow_measurement(model=model)
 print(f"Estimated expectation value shadow = {estimated_values_shadow}") # markdown-exec: hide
 ```
 
+Note that the optional `n_shots` is by default 1, which means for one unitary, we sample only once. If we specify a higher
+value, more shots are realized per sampled unitary, and a different formula is used involving the Hamming distance denoted $D$ (see Eq. 2.42 of Ref[^3]):
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]:
```suggestion
Note that the optional `n_shots` is by default 1, which means for one unitary, we sample only once. If we specify a higher
number of shots, more samples are realized per unitary accordingly, and a different formula is used involving the Hamming distance denoted $D$ (see Eq. 2.42 of Ref[^3]):
```

File: docs/measurements/index.md (hunk 1 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md]: Yeah basically is it an option that you have to pass here or a full-fledged argument to the `measure` function ?

File: qadence_protocols/measurements/shadow.py [hunk 3]
```diff
@@ -201,9 +225,13 @@ def expectation(
             if len(observables) > 0
             else [obs.abstract for obs in self.model._observable]
         )
-        _, K = number_of_samples(observables=observables, accuracy=accuracy, confidence=confidence)
+        _, K = number_of_samples(
+            observables=observables, accuracy=accuracy or 0.0, confidence=confidence or 0
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py]: Wouldn't it be better if the default gets set using `.get` ? Just like the number of shots above ?

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    Compute local shadow by inverting the quantum channel for each projector state.
+
+    See https://arxiv.org/pdf/2002.08953.pdf
+    Supplementary Material 1 and Eqs. (S17,S44).
+
+    Expects a sample bitstring in ILO.
+    """
+
+    nested_unitaries = rotations_unitary_map(unitary_ids)
+    nested_unitaries_adjoint = rotations_unitary_map(unitary_ids, UNITARY_TENSOR_adjoint)
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py]:
```suggestion
    nested_unitaries_adjoint = rotations_unitary_map(unitary_ids, UNITARY_TENSOR_ADJOINT)
```

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        bitstrings.bool().unsqueeze(-1).unsqueeze(-1), P1_MATRIX, P0_MATRIX
+    )
+    idmatcal = torch.stack([idmat * 0.5 * (1.0 / corr_coeff - 1.0) for corr_coeff in calibration])
+    local_densities = (1.0 / calibration.unsqueeze(-1).unsqueeze(-1)) * (
+        nested_unitaries_adjoint @ projmat @ nested_unitaries
+    ) - idmatcal
+    return local_densities
+
+
+def global_shadow_Hamming(probas: Tensor, unitary_ids: Tensor) -> Tensor:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py]:
```suggestion
def global_shadow_hamming(probas: Tensor, unitary_ids: Tensor) -> Tensor:
```

File: qadence_protocols/measurements/utils_shadow/post_processing.py (hunk 4 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/utils_shadow/post_processing.py]: Another option would be to pass only a boolean here (like for instance `dagger = False`) and let `rotation_unitary_map` resolve the correct matrices.

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+    See https://arxiv.org/pdf/2002.08953.pdf
+    Supplementary Material 1 and Eqs. (S17,S44).
+
+    Expects a sample bitstring in ILO.
+    """
+
+    nested_unitaries = rotations_unitary_map(unitary_ids)
+    nested_unitaries_adjoint = rotations_unitary_map(unitary_ids, UNITARY_TENSOR_adjoint)
+    projmat = torch.empty(nested_unitaries.shape, dtype=nested_unitaries.dtype)
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py]:
```suggestion
    proj_mat = torch.empty(nested_unitaries.shape, dtype=nested_unitaries.dtype)
```

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+    projmat[..., :, :] = torch.where(
+        bitstrings.bool().unsqueeze(-1).unsqueeze(-1), P1_MATRIX, P0_MATRIX
+    )
+    local_densities = 3.0 * (nested_unitaries_adjoint @ projmat @ nested_unitaries) - idmat
+    return local_densities
+
+
+def robust_local_shadow(bitstrings: Tensor, unitary_ids: Tensor, calibration: Tensor) -> Tensor:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py:63]: There is quite an overlap with the function above. Is there any way to make functionality reusable ?

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    local_densities = (1.0 / calibration.unsqueeze(-1).unsqueeze(-1)) * (
+        nested_unitaries_adjoint @ projmat @ nested_unitaries
+    ) - idmatcal
+    return local_densities
+
+
+def global_shadow_Hamming(probas: Tensor, unitary_ids: Tensor) -> Tensor:
+    """Compute global shadow using a Hamming matrix."""
+
+    nested_unitaries = rotations_unitary_map(unitary_ids)
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py:76]: Just wondering: since they are common to at least 3 methods, does it make sense to import them module-wide ?

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+    nested_unitaries_adjoint = rotations_unitary_map(unitary_ids, UNITARY_TENSOR_adjoint)
+
+    N = unitary_ids.shape[1]
+    if N > 1:
+        nested_unitaries = batch_kron(nested_unitaries)
+        nested_unitaries_adjoint = batch_kron(nested_unitaries_adjoint)
+    Hamming_mat = [HammingMatrix.to(dtype=probas.dtype) for i in range(N)]
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py]:
```suggestion
    hamming_mat = [HammingMatrix.to(dtype=probas.dtype) for i in range(N)]
```

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    ein_command += "->" + einsum_alphabet[0] + einsum_alphabet_cap[1 : N + 1]
+    probprime = d * torch.einsum(ein_command, *([probas] + Hamming_mat)).to(
+        dtype=nested_unitaries.dtype
+    )
+    probprime = torch.diag_embed(probprime.reshape((-1, d)))
+    densities = nested_unitaries_adjoint @ probprime @ nested_unitaries
+    return densities
+
+
+def global_robust_shadow_Hamming(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py]: Same. It feels like there is overlapping functionality that can be grouped together.

File: qadence_protocols/measurements/utils_shadow/post_processing.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+    snapshots: list = list()
+
+    if local_shadows and unitaries_ids.shape[1] > 1:
+
+        def obtain_global_shadow(bits: Tensor, unit_ids: Tensor) -> Tensor:
+            return batch_kron(shadow_caller(bits, unit_ids))
+
+    else:
+
+        def obtain_global_shadow(bits: Tensor, unit_ids: Tensor) -> Tensor:
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow/post_processing.py:158]: Is this necessary ? Wouldn't `batch_kron` handle properly the case when there is a unique global shadow ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. I think there can be some general grouping of functionality that would make the whole code more readable and reusable. Also, I thought the number of shots was passed in the measure method ? It feels a but weird that it is passed as an option unless there is smtg I don't see now.

File: docs/measurements/index.md (hunk 1 above)
Comment (chMoussa) [reply docs/measurements/index.md]: I think here because n_shots can be used to validate the data passed to a protocol.

File: qadence_protocols/measurements/utils_shadow/post_processing.py (hunk 4 above)
Comment (chMoussa) [reply qadence_protocols/measurements/utils_shadow/post_processing.py]: I wanted to keep this function general for now, say we introduce other unitaries.

File: qadence_protocols/measurements/utils_shadow/post_processing.py (hunk 11 above)
Comment (chMoussa) [reply qadence_protocols/measurements/utils_shadow/post_processing.py:158]: Maybe but that avoid useless `batch_kron` calls in one if

**Comment by chMoussa** [comment]
Hey @RolandMacDoland , indeed it was possible to refactor a few instructions. I still kept different functions as it avoids rerunning many if statements for when using global or local shadows, and using robust or vanilla shadows.

File: docs/measurements/index.md (hunk 1 above)
Comment (RolandMacDoland) [reply docs/measurements/index.md]: hm, not too sure about it.

File: qadence_protocols/measurements/shadow.py [hunk 12]
```diff
@@ -57,17 +62,20 @@ def validate_options(self, options: dict) -> dict:
         """
 
         shadow_size = options.get("shadow_size", None)
-        accuracy = options.get("accuracy", None)
-        if shadow_size is None and accuracy is None:
+        accuracy = options.get("accuracy", 0)
+        if shadow_size is None and accuracy == 0:
             raise KeyError(
                 "Shadow protocol requires either an option"
                 " 'shadow_size' of type 'int' or 'accuracy' of type 'float'."
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py]: I guess it feels a bit contradictory with the error message because `0` can still be considered a `float` (or `int`). I would keep `None` as it makes more sense IMHO.

File: qadence_protocols/measurements/utils_shadow/post_processing.py (hunk 11 above)
Comment (RolandMacDoland) [reply qadence_protocols/measurements/utils_shadow/post_processing.py:158]: Yeah but the penalty is negligible. I guess you gain more by avoiding the conditionals if possible.

**Review by RolandMacDoland — APPROVED** [review]
Hiya, looks good to me. Thanks. SOme more minor quibbles.
